### ISSUE TITLE
Packing in SFT

### ISSUE THREAD

**Issue by wdykas** [huggingface/trl#805]
I understand how packing is allowed in pretraining but I was looking for some clarification on how we are allowed to pack samples for SFT with `ConstantLengthDataset`.   I see that an EOS token is put between samples https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L567 but how are attention masks handled to make sure that the samples don't attend to each others context or is just putting EOS enough signal?

Could you also point me to the code for how multiple answers are handled in generation for loss calculations?

**Comment by younesbelkada** [comment]
Hi @wdykas !

Packing is a common practice and a trick to enable pre-training / fine-tuning on more sequences. I think that adding the EOS token is an enough signal for the model. 

![Screenshot 2023-09-25 at 14 58 14](https://github.com/huggingface/trl/assets/49240599/d9a7eea4-acb1-4cf4-a44f-f0c9be28175e)

The trick seems to has been introduced in the T5 paper: https://arxiv.org/pdf/1910.10683.pdf

> Could you also point me to the code for how multiple answers are handled in generation for loss calculations?

Can you elaborate more on this question? for loss calculation we do not perform generation we only compute the next token once.

**Comment by wdykas** [comment]
Sorry I meant more along the lines of how can calculate loss on the tokens without cross contamination in attention between sequences?  Is there code where you create special attention masks? This seemed to be an issue opened here: https://github.com/huggingface/transformers/issues/25452

**Comment by lvwerra** [comment]
Since the sequences are generally not correlated there is no issue in cross contamination. Otherwise you would need to pass a dedicated attention mask but this is not currently done.

**Comment by wdykas** [comment]
thanks for the answer! One last QQ, a problem that I see in this implementation is that we create a long continuous buffer of text separated by [EOS]. Then it seems we create examples of seq_len by naively looping over https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L579C35-L579C48 without considering EOS.
 Doesn't this mean that some examples, if not many, will start with sequences that are cutoff and are in the middle?

**Comment by lvwerra** [comment]
Yes, that's the idea of packing. If you want to avoid that you can disable packing and each sequence will start at the beginning and filled with padding tokens to the max sequence length in the batch.

**Comment by btrcm00** [comment]
> ![image](https://user-images.githubusercontent.com/26831266/272305004-93c690a8-7e9b-40ad-885f-d530996aa109.png)
> 
> I think it is better to add a feature to make the sequence packing without any kinds of cross contamination like the figure above. Specifically in the supervised finetuning stage. Because in the SFT stage, We usually have short conversations (e.g. << 4096), which may cause some "training-inference gap" if we don't add the attention mask.
> 
> Notice that there is an implicit affect of the very long sequence attention (if we do not add the attention mask) -- the entropy [1].
> 
> [1] https://kexue.fm/archives/8823

Thank you for the answer!

**Comment by lvwerra** [comment]
Packing is mainly a throughput/efficiency technique at scale. For 1000 samples, why don't you just turn packing off?

**Comment by Qubitium** [comment]
> Packing is mainly a throughput/efficiency technique at scale. For 1000 samples, why don't you just turn packing off?

Because gpu waste is waste at any scale. The idea of packing allows us to reduce small/micro training time to be completed in 1/2 of the time vs packing off. Imagine 1 minute training time (packing=on) vs 2 minutes (packing=off) but repeat that N times over the course of the day for a developer that is trying to maximize work/time efficiency.  We have verified the training time diff for packing=on but result quality is an issue for us. 

 I don't think packing should be reserved for large data. For many cases, 1000 dataset for finetuning is actually overkill.  Quality of data trumps dataset row count. So 1k may be tiny, but actually for a special laser focused training, it's quite large. 

Just that the current implementations is not optimized for correlated/small datasets, and/or that we are not using it properly, thus my questions on the two points that is causing confusion/problem for us.

**Comment by hanyin88** [comment]
Thanks for sharing this very important topic. 

Another likely very relevant consideration is that, in Huggingface's BetterTransformer (Fast Kernel in LLaMA-recipes) does not support attention_mask:

> The PyTorch-native scaled_dot_product_attention operator can only dispatch to Flash Attention if no attention_mask is provided. Thus, by default in training mode, the BetterTransformer integration drops the mask support and can only be used for training that do not require a padding mask for batched training. BetterTransformer is not suited for the fine-tuning of models on tasks that requires a padding mask.

**Comment by khai-meetkai** [comment]
You can use the **correct implementation of packing without cross-contamination attention** from this repo: https://github.com/MeetKai/functionary/tree/main/functionary/train/packing

**Comment by gotzmann** [comment]
@khai-meetkai will it work with Unsloth? https://github.com/unslothai/unsloth

**Comment by cassanof** [comment]
> You can use the **correct implementation of packing without cross-contamination attention** from this repo: https://github.com/MeetKai/functionary/tree/main/functionary/train/packing

Not correct. you are using padding....

**Comment by Qubitium** [comment]
> > You can use the **correct implementation of packing without cross-contamination attention** from this repo: https://github.com/MeetKai/functionary/tree/main/functionary/train/packing
> 
> Not correct. you are using padding....

It is padding packed entries that is shorter than max. Or are you referring to something else? Packing does not mean you can't use padding.

**Comment by cassanof** [comment]
> > > You can use the **correct implementation of packing without cross-contamination attention** from this repo: https://github.com/MeetKai/functionary/tree/main/functionary/train/packing
> > 
> > 
> > Not correct. you are using padding....
> 
> It is padding packed entries that is shorter than max. Or are you referring to something else? Packing does not mean you can't use padding.

I think this thread is about the ConstantLengthDatasetLoader, but correct me if I'm wrong. One of the advantages of using CLDL is that you don't use any padding, thus training could be more stable in some circumstances / downstream tasks. Yes, you can pack with padding too, but it wouldn't be the same thing,

**Comment by Qubitium** [comment]
> I think this thread is about the ConstantLengthDatasetLoader, but correct me if I'm wrong. One of the advantages of using CLDL is that you don't use any padding, thus training could be more stable in some circumstances / downstream tasks. Yes, you can pack with padding too, but it wouldn't be the same thing,

Ok I see your point. I would argue that  constantlength with no padding which naively splits just to fill max length is throwing some daset rows out the window. Now if your train set is millions of rows, this may not matter but smaller datasets, you most definitely want to pack as much into max len with no naive splitting and pad the rest.

Also can you point to papers or articles showing padding in general affecting quality/stability of training? I have never trained with no padding so I have no experience in this regard.

**Comment by Peter-Devine** [comment]
> you would need to pass a dedicated attention mask but this is not currently done

Is this done now? I'd like to know if TRL's implementation attends to all samples at once. I know packages such as [Axolotl](https://github.com/OpenAccess-AI-Collective/axolotl/blob/00018629e710ef8166c3834d7934d19088a740a8/docs/multipack.md) support this.

Thanks!

**Comment by yuleiqin** [comment]
Anything update? I am also interested to use packing for instruction finetuning on long-context examples.

In that case, I would combine both short and long context examples. To improve training efficiency, I think packing is useful to maximize your training efficiency (GPU memory) for each batch. However, the current implementation of packing might not be optimum for the SFT datasets:
1) each sample in one concatenated overall example should not attend to each other.
2) each sample should not be cut in the middle because the answer should be following the instruction without any loss of information.

Anything new for the solution? Thanks in advance for your help.
